contracts/EtomicSwap.sol

```diff
@@ -115,25 +279,143 @@ contract EtomicSwap {
                 _tokenAddress,
                 _amount
-            ));
+            )
+        );
 
-        require(paymentHash == payments[_id].paymentHash && now >= payments[_id].lockTime);
+        require(
+            paymentHash == payments[_id].paymentHash &&
+            block.timestamp >= payments[_id].lockTime
+        );
 
         payments[_id].state = PaymentState.SenderRefunded;
 
         if (_tokenAddress == address(0)) {
-            msg.sender.transfer(_amount);
+            payable(msg.sender).transfer(_amount);
         } else {
             IERC20 token = IERC20(_tokenAddress);
             require(token.transfer(msg.sender, _amount));
         }
 
         emit SenderRefunded(_id);
     }
+
+    function senderRefundErc721(
+        bytes32 _id,
+        bytes20 _paymentHash,
+        address _tokenAddress,
+        uint256 _tokenId,
+        address _receiver
+    ) external {
+        require(payments[_id].state == PaymentState.PaymentSent);
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                _receiver,
+                msg.sender,
+                _paymentHash,
+                _tokenAddress,
+                _tokenId
+            )
+        );
+
+        require(
+            paymentHash == payments[_id].paymentHash &&
+            block.timestamp >= payments[_id].lockTime
+        );
+        payments[_id].state = PaymentState.SenderRefunded;
+
+        IERC721 token = IERC721(_tokenAddress);
+        token.safeTransferFrom(address(this), msg.sender, _tokenId);
+
+        emit SenderRefunded(_id);
+    }
+
+    function senderRefundErc1155(
+        bytes32 _id,
+        uint256 _amount,
+        bytes20 _paymentHash,
+        address _tokenAddress,
+        uint256 _tokenId,
+        address _receiver
+    ) external {
+        require(payments[_id].state == PaymentState.PaymentSent);
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                _receiver,
+                msg.sender,
+                _paymentHash,
+                _tokenAddress,
+                _tokenId,
+                _amount
+            )
+        );
+
+        require(
+            paymentHash == payments[_id].paymentHash &&
+            block.timestamp >= payments[_id].lockTime
+        );
+        payments[_id].state = PaymentState.SenderRefunded;
+
+        IERC1155 token = IERC1155(_tokenAddress);
+        token.safeTransferFrom(
+            address(this),
+            msg.sender,
+            _tokenId,
+            _amount,
+            ""
+        );
+
+        emit SenderRefunded(_id);
+    }
+
+    function onERC1155Received(
+        address, /* operator */
+        address, /* from */
+        uint256, /* id */
+        uint256, /* value */
+        bytes calldata /* data */
+    ) external pure override returns (bytes4) {
+        // Return this magic value to confirm receipt of ERC1155 token
+        return this.onERC1155Received.selector;
```

Comment:
This means that a user can transfer ERC1155 directly to the etomic swap smart contract address, and token(s) will be actually burnt, as there is no way to transfer it out of the contract without creating HTLC payment. So I think that we even should handle payment creation logic here (see other comment).

Reply:
added HTLC logic into onERC1155Received